composer/metrics/map.py

```diff
@@ -0,0 +1,383 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+# Adapted from https://github.com/Lightning-AI/metrics/blob/1c42f6643f9241089e55a4d899f14da480021e16/torchmetrics/detection/map.py
+# Current versions of MAP in torchmetrics are incorrect or slow as of 9/21/22.
+# Relevant issues:
+# https://github.com/Lightning-AI/metrics/issues/1024
+# https://github.com/Lightning-AI/metrics/issues/1164
+
+"""MAP torchmetric for object detection."""
+#type: ignore
+import logging
+import sys
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, List, Optional, Sequence, Union
+
+import torch
+from torch import Tensor
+from torchmetrics.metric import Metric
+from torchvision.ops import box_convert
+
+from composer.utils.import_helpers import MissingConditionalImportError
+
+__all__ = ['MAP']
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class MAPMetricResults:
+    """Dataclass to wrap the final mAP results."""
+    map: Tensor
+    map_50: Tensor
+    map_75: Tensor
+    map_small: Tensor
+    map_medium: Tensor
+    map_large: Tensor
+    mar_1: Tensor
+    mar_10: Tensor
+    mar_100: Tensor
+    mar_small: Tensor
+    mar_medium: Tensor
+    mar_large: Tensor
+    map_per_class: Tensor
+    mar_100_per_class: Tensor
+
+    def __getitem__(self, key: str) -> Union[Tensor, List[Tensor]]:
+        return getattr(self, key)
+
+
+# noinspection PyMethodMayBeStatic
+class WriteToLog:
+    """Logging class to move logs to log.debug()."""
+
+    def write(self, buf: str) -> None:  # skipcq: PY-D0003, PYL-R0201
+        for line in buf.rstrip().splitlines():
+            log.debug(line.rstrip())
+
+    def flush(self) -> None:  # skipcq: PY-D0003, PYL-R0201
+        for handler in log.handlers:
+            handler.flush()
+
+    def close(self) -> None:  # skipcq: PY-D0003, PYL-R0201
+        for handler in log.handlers:
+            handler.close()
+
+
+class _hide_prints:
+    """Internal helper context to suppress the default output of the pycocotools package."""
+
+    def __init__(self) -> None:
+        self._original_stdout = None
+
+    def __enter__(self) -> None:
+        self._original_stdout = sys.stdout  # type: ignore
+        sys.stdout = WriteToLog()  # type: ignore
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:  # type: ignore
+        sys.stdout.close()
+        sys.stdout = self._original_stdout  # type: ignore
+
+
+def _input_validator(preds: List[Dict[str, torch.Tensor]], targets: List[Dict[str, torch.Tensor]]) -> None:
```

Comment:
yes I did pull it from the original and it does looks a bit crazy, but it shouldn't be too much iteration. Each `pred` or `target` list will be at most `batch_size` long. batches don't get very large in object detection land, even 64 is sizable. There will be a `box`, `score`, and `label` for each detection in each image so anywhere from 0 to ~200 for something like coco. I was worried about the last iteration for this,  but since it's just checking a dimension of the `boxes`, `labels` and `scores` tensors that's a constant time lookup rather than an iteration through every object. So still no more than `batch_size` iters.

Reply:
Thanks for the clarification!